docs/source/en/llm_optims.md

```diff
@@ -0,0 +1,271 @@
+
+Basic inference is slow because LLMs are autoregressive, which means it becomes progressively slower the more tokens it generates. Additionally, LLMs have billions of parameters, making it a challenge to store and handle all those weights in memory. However, with a few optimization techniques, you can boost your inference speed and reduce memory usage.
+
+This guide will show you how to use the optimization techniques available in Transformers to accelerate LLM inference.
+
+> [!TIP]
+> Hugging Face also provides [Text Generation Inference (TGI)](https://hf.co/docs/text-generation-inference), a library dedicated to deploying and serving highly optimized LLMs for inference. It includes more optimization features not included in Transformers such as continuous batching for increasing throughput and tensor parallelism for multi-GPU inference.
+
+## Static kv-cache and torch.compile
+
+During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time since the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.
+
+To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses Python code into fast and optimized kernels.
+
+The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
+
+> [!WARNING]
+> Currently, only [Gemma](./model_doc/gemma) and [Llama](./model_doc/llama2) models support static kv-cache and torch.compile.
+
+For this example, let's load the [Gemma](https://hf.co/google/gemma-2b) model.
+
+```py
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
+
+model = AutoModelForCausalLM.from_pretrained(
+    "google/gemma-2b", device_map="auto"
+)
+```
+
+There are two ways you can configure the model to use a static kv-cache.
+
+<hfoptions id="static-kv">
+<hfoption id="generation_config">
+
+Access the model's `generation_config` attribute and set the `cache_implementation` to "static".
+
+```py
+model.generation_config.cache_implementation = "static"
+```
+
+Call torch.compile on the model to compile it with the static kv-cache.
+
+```py
+compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
+input_text = "Planck's constant is"
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+
+outputs = compiled_model.generate(**input_ids)
```

Comment:
generate is not yet compatible with torch compile so would only recommend the forward pass to be shown here! And talk about 4x!